src/expression.rs

```diff
@@ -206,4 +222,12 @@ mod tests {
         assert!(parse_num("+6").is_err());
         assert!(parse_num("-6").is_err());
     }
+
+    #[test]
+    #[ignore]
```

Comment:
This `#[ignore]` can be removed. (The `heavy_nest` one cannot.)

Reply:
Why? are they not the same thing?

Reply:
The `heavy_nest` one stack overflows on my system while this one does not.

I assumed the other stack overflow would be fixed in a future PR.

Reply:
Interesting, does it show an overflow error? Or just take a long time because there is no recursion here.

Reply:
The reason for #[ignore] is that all ignored tests are run in release mode in CI. In debug mode, the rustc keeps lot of debug information which makes the program really slow to execute. 

In the first commit of the PR, both of the tests should pass in a few seconds in release mode. there is no bug fix for a future commit

Reply:
Based on a discussion in IRC, I have reverted the commits that switched the implementation back to recursive ones.